File: python/engine.py
import asyncio
import time
import uuid
from abc import ABC, abstractmethod

from models import DecomposerResponse, HydraNode, NodeMetadata


class Adapter(ABC):
    @abstractmethod
    async def decompose(self, topic: str, breadth: int, system_prompt: str = None) -> DecomposerResponse:
        pass

    @abstractmethod
    def get_model_name(self) -> str:
        pass


class HydraConfig:
    def __init__(
        self,
        depth_limit: int,
        branching_factor: int,
        adapter: Adapter,
        persona: str = None,
        system_prompt: str = None,
    ):
        if persona and system_prompt:
            raise ValueError("Cannot specify both persona and system_prompt")
        self.depth_limit = depth_limit
        self.branching_factor = branching_factor
        self.adapter = adapter
        self.persona = persona
        self.system_prompt = system_prompt


class HydraEngine:
    def __init__(self, config: HydraConfig):
        self.config = config
        self.resolved_system_prompt = config.system_prompt or config.persona

    async def run(self, prompt: str) -> HydraNode:
        return await self._expand(prompt, 0)

    async def _expand(self, topic: str, current_depth: int) -> HydraNode:
        node = HydraNode(
            id=str(uuid.uuid4()),
            topic=topic,
            depth=current_depth,
            metadata=NodeMetadata(
                tokens=0, model=self.config.adapter.get_model_name(), latency_ms=0
            ),
            children=[],
        )

        if current_depth >= self.config.depth_limit:
            node.status = "success"
            return node

        try:
            start_time = time.time()
            response = await self.config.adapter.decompose(
                topic, self.config.branching_factor, self.resolved_system_prompt
            )
            end_time = time.time()

            node.metadata.tokens = response.metadata.tokens
            node.metadata.latency_ms = int((end_time - start_time) * 1000)

            # FR-2: Parallel Execution with asyncio.gather
            child_tasks = [
                self._expand(subtopic, current_depth + 1) for subtopic in response.subtopics
            ]

            node.children = list(await asyncio.gather(*child_tasks))
            node.status = "success"
        except Exception:
            node.status = "failed"
            # Optional: print(f"Error expanding {topic}: {e}")

        return node

File: go/engine.go
package hydra

import (
	"sync"
	"time"

	"github.com/google/uuid"
)

type HydraConfig struct {
	DepthLimit      int
	BranchingFactor int
	Adapter         Adapter
	Persona         string // "analytical" or "action"
	SystemPrompt    string // custom prompt (overrides Persona)
}

type HydraEngine struct {
	config               HydraConfig
	resolvedSystemPrompt string
}

func NewHydraEngine(config HydraConfig) *HydraEngine {
	// Validate: can't have both
	if config.Persona != "" && config.SystemPrompt != "" {
		panic("cannot specify both Persona and SystemPrompt")
	}
	resolved := config.SystemPrompt
	if resolved == "" && config.Persona != "" {
		resolved = config.Persona
	}
	return &HydraEngine{
		config:               config,
		resolvedSystemPrompt: resolved,
	}
}

func (e *HydraEngine) Run(prompt string) (*HydraNode, error) {
	return e.expand(prompt, 0), nil
}

func (e *HydraEngine) expand(topic string, currentDepth int) *HydraNode {
	node := &HydraNode{
		ID:    uuid.New().String(),
		Topic: topic,
		Depth: currentDepth,
		Metadata: NodeMetadata{
			Model: e.config.Adapter.GetModelName(),
		},
		Children: []*HydraNode{},
	}

	if currentDepth >= e.config.DepthLimit {
		node.Status = "success"
		return node
	}

	start := time.Now()
	response, err := e.config.Adapter.Decompose(topic, e.config.BranchingFactor, e.resolvedSystemPrompt)
	latency := int(time.Since(start).Milliseconds())

	if err != nil {
		node.Status = "failed"
		return node
	}

	node.Metadata.Tokens = response.Metadata.Tokens
	node.Metadata.LatencyMS = latency

	var wg sync.WaitGroup
	childChan := make(chan *HydraNode, len(response.Subtopics))

	for _, subtopic := range response.Subtopics {
		wg.Add(1)
		go func(st string) {
			defer wg.Done()
			childChan <- e.expand(st, currentDepth+1)
		}(subtopic)
	}

	// Wait for all branches in parallel
	wg.Wait()
	close(childChan)

	for child := range childChan {
		node.Children = append(node.Children, child)
	}

	node.Status = "success"
	return node
}
